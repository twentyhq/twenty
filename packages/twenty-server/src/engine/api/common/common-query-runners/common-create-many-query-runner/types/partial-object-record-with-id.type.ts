import { type ObjectRecord } from 'twenty-shared/types';

export type PartialObjectRecordWithId = Partial<ObjectRecord> & { id: string };
