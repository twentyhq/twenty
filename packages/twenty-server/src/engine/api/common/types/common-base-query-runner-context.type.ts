import { type ObjectLiteral } from 'typeorm';

import { type CommonQueryRunnerOptions } from 'src/engine/api/common/interfaces/common-query-runner-options.interface';
import { type ResolverArgs } from 'src/engine/api/graphql/workspace-resolver-builder/interfaces/workspace-resolvers-builder.interface';

import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';

export type CommonBaseQueryRunnerContext<Input extends ResolverArgs> = {
  args: Input;
  options: CommonQueryRunnerOptions;
  workspaceDataSource: WorkspaceDataSource;
  repository: WorkspaceRepository<ObjectLiteral>;
  selectedFieldsResult: CommonSelectedFieldsResult;
  isExecutedByApiKey: boolean;
  roleId?: string;
  shouldBypassPermissionChecks: boolean;
};
