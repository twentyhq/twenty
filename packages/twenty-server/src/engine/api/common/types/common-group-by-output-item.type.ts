type AggregateValues = {
  [key: string]: number;
};

type GroupByDimensionValues = {
  groupByDimensionValues: string[];
};

export type CommonGroupByOutputItem = AggregateValues & GroupByDimensionValues;
