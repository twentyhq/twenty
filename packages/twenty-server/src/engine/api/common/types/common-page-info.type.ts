import { type PageInfo } from 'src/engine/api/rest/core/interfaces/rest-api-base.handler';

export type CommonPageInfo = {
  hasNextPage: NonNullable<PageInfo['hasNextPage']>;
  hasPreviousPage: NonNullable<PageInfo['hasPreviousPage']>;
  startCursor: string | null;
  endCursor: string | null;
};
