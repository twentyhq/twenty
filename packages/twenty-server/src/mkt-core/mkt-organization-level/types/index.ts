export * from './organization-level.types';
export * from './permissions.type';
