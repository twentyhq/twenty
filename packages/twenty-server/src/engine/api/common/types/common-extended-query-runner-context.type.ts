import { type ObjectLiteral } from 'typeorm';

import { type WorkspaceAuthContext } from 'src/engine/api/common/interfaces/workspace-auth-context.interface';

import { type CommonBaseQueryRunnerContext } from 'src/engine/api/common/types/common-base-query-runner-context.type';
import { type GraphqlQueryParser } from 'src/engine/api/graphql/graphql-query-runner/graphql-query-parsers/graphql-query.parser';
import { type WorkspaceDataSource } from 'src/engine/twenty-orm/datasource/workspace.datasource';
import { type WorkspaceRepository } from 'src/engine/twenty-orm/repository/workspace.repository';
import { type RolePermissionConfig } from 'src/engine/twenty-orm/types/role-permission-config';

export type CommonExtendedQueryRunnerContext = Omit<
  CommonBaseQueryRunnerContext,
  'authContext'
> & {
  authContext: WorkspaceAuthContext;
  rolePermissionConfig: RolePermissionConfig;
  repository: WorkspaceRepository<ObjectLiteral>;
  commonQueryParser: GraphqlQueryParser;
  workspaceDataSource: WorkspaceDataSource;
};
