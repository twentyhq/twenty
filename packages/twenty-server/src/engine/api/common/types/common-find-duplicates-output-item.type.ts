import { type ObjectRecord } from 'twenty-shared/types';

export type CommonFindDuplicatesOutputItem = {
  records: ObjectRecord[];
  totalCount: number;
  hasNextPage: boolean;
  hasPreviousPage: boolean;
  startCursor: string | null;
  endCursor: string | null;
};
