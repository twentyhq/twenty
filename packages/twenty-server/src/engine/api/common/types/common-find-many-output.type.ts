import { type ObjectRecord } from 'twenty-shared/types';

import { type CommonPageInfo } from 'src/engine/api/common/types/common-page-info.type';
import { type CommonSelectedFieldsResult } from 'src/engine/api/common/types/common-selected-fields-result.type';

export type CommonFindManyOutput = {
  records: ObjectRecord[];
  aggregatedValues: Record<string, number>;
  totalCount: number;
  pageInfo: CommonPageInfo;
  selectedFieldsResult: CommonSelectedFieldsResult;
};
