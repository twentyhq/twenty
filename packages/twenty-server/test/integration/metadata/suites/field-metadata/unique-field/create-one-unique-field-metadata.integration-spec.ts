import { createOneFieldMetadata } from 'test/integration/metadata/suites/field-metadata/utils/create-one-field-metadata.util';
import { createOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/create-one-object-metadata.util';
import { deleteOneObjectMetadata } from 'test/integration/metadata/suites/object-metadata/utils/delete-one-object-metadata.util';
import { FieldMetadataType } from 'twenty-shared/types';

describe('create one unique field metadata', () => {
  let createdObjectMetadataId = '';

  beforeEach(async () => {
    const {
      data: {
        createOneObject: { id: objectMetadataId },
      },
    } = await createOneObjectMetadata({
      input: {
        nameSingular: 'testObject',
        namePlural: 'testObjects',
        labelSingular: 'Test Object',
        labelPlural: 'Test Objects',
        icon: 'IconTest',
      },
    });

    createdObjectMetadataId = objectMetadataId;
  });

  afterEach(async () => {
    await deleteOneObjectMetadata({
      input: { idToDelete: createdObjectMetadataId },
    });
  });

  it('should create one unique field metadata', async () => {
    const createFieldInput = {
      name: 'uniqueTestField',
      label: 'Unique Test Field',
      type: FieldMetadataType.TEXT,
      objectMetadataId: createdObjectMetadataId,
      isUnique: true,
    };

    const { data, errors } = await createOneFieldMetadata({
      input: createFieldInput,
      gqlFields: `
        id
        name
        label
        type
        isUnique
      `,
    });

    expect(errors).toBeUndefined();
    expect(data).not.toBeNull();
    expect(data.createOneField).toBeDefined();
    expect(data.createOneField.name).toBe('uniqueTestField');
    expect(data.createOneField.isUnique).toBe(true);
  });

  it('should not create unique field metadata if it has custom default value', async () => {
    const createFieldInput = {
      name: 'uniqueFieldWithDefault',
      label: 'Unique Field With Default',
      type: FieldMetadataType.TEXT,
      objectMetadataId: createdObjectMetadataId,
      isUnique: true,
      defaultValue: "'?'",
    };

    const { data, errors } = await createOneFieldMetadata({
      input: createFieldInput,
      gqlFields: `
        id
        name
        label
        type
        isUnique
      `,
      expectToFail: true,
    });

    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors[0].message).toBe('Unique field cannot have a default value');
  });

  it('should not create unique field metadata for composite type without unique subfields', async () => {
    const createFieldInput = {
      name: 'uniqueFullName',
      label: 'Unique Full Name',
      type: FieldMetadataType.FULL_NAME,
      objectMetadataId: createdObjectMetadataId,
      isUnique: true,
    };

    const { data, errors } = await createOneFieldMetadata({
      input: createFieldInput,
      gqlFields: `
        id
        name
        label
        type
        isUnique
      `,
      expectToFail: true,
    });

    expect(data).toBeNull();
    expect(errors).toBeDefined();
    expect(errors[0].message).toContain(
      'Unique index cannot be created for field uniqueFullName of type FULL_NAME',
    );
  });

  it('should create unique field metadata for composite type with unique subfields', async () => {
    const createFieldInput = {
      name: 'uniqueEmails',
      label: 'Unique Emails',
      type: FieldMetadataType.EMAILS,
      objectMetadataId: createdObjectMetadataId,
      isUnique: true,
    };

    const { data, errors } = await createOneFieldMetadata({
      input: createFieldInput,
      gqlFields: `
        id
        name
        label
        type
        isUnique
      `,
    });

    expect(errors).toBeUndefined();
    expect(data).not.toBeNull();
    expect(data.createOneField).toBeDefined();
    expect(data.createOneField.name).toBe('uniqueEmails');
    expect(data.createOneField.isUnique).toBe(true);
  });
});
