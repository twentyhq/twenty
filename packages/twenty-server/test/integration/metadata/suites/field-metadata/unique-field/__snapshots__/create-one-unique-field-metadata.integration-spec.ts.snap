// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`failing createOne unique field metadata should fail for composite type without unique subfields (FULL_NAME) 1`] = `
{
  "extensions": {
    "code": "METADATA_VALIDATION_FAILED",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [
        {
          "errors": [
            {
              "code": "INDEX_FIELD_INVALID_TYPE_FOR_UNIQUE",
              "message": "Unique index cannot be created for field uniqueFullName of type FULL_NAME",
              "userFriendlyMessage": "FULL_NAME fields cannot be unique.",
            },
          ],
          "flatEntityMinimalInformation": {
            "id": Any<String>,
            "name": "IDX_UNIQUE_98d410e804d758f9bcceb154dde",
          },
          "status": "fail",
          "type": "create_index",
        },
      ],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while creating fields",
  "name": "GraphQLError",
}
`;

exports[`failing createOne unique field metadata should fail when field has custom default value 1`] = `
{
  "extensions": {
    "code": "METADATA_VALIDATION_FAILED",
    "errors": {
      "cronTrigger": [],
      "databaseEventTrigger": [],
      "fieldMetadata": [],
      "index": [
        {
          "errors": [
            {
              "code": "INDEX_FIELD_INVALID_DEFAULT_VALUE",
              "message": "Unique index cannot be created for field uniqueFieldWithDefault of type TEXT",
              "userFriendlyMessage": "TEXT fields cannot have a default value.",
            },
          ],
          "flatEntityMinimalInformation": {
            "id": Any<String>,
            "name": "IDX_UNIQUE_a1b02809c523cdc853d35fbc413",
          },
          "status": "fail",
          "type": "create_index",
        },
      ],
      "objectMetadata": [],
      "routeTrigger": [],
      "serverlessFunction": [],
      "view": [],
      "viewField": [],
      "viewFilter": [],
      "viewGroup": [],
    },
    "message": "Validation failed for 0 object(s) and 0 field(s)",
    "summary": {
      "invalidCronTrigger": 0,
      "invalidDatabaseEventTrigger": 0,
      "invalidFieldMetadata": 0,
      "invalidIndex": 0,
      "invalidObjectMetadata": 0,
      "invalidRouteTrigger": 0,
      "invalidServerlessFunction": 0,
      "invalidView": 0,
      "invalidViewField": 0,
      "invalidViewFilter": 0,
      "invalidViewGroup": 0,
      "totalErrors": 0,
    },
    "userFriendlyMessage": "Validation failed for 0 object(s) and 0 field(s)",
  },
  "message": "Multiple validation errors occurred while creating fields",
  "name": "GraphQLError",
}
`;
